sdk/identity/azure-identity/inc/azure/identity/workload_identity_credential.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * @file
+ * @brief Workload Identity Credential and options.
+ */
+
+#pragma once
+
+#include "azure/identity/detail/client_credential_core.hpp"
+#include "azure/identity/detail/token_cache.hpp"
+
+#include <azure/core/credentials/token_credential_options.hpp>
+
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Identity {
+  namespace _detail {
+    class TokenCredentialImpl;
+  } // namespace _detail
+
+  /**
+   * @brief Options for workload identity credential.
+   *
+   */
+  struct WorkloadIdentityCredentialOptions final : public Core::Credentials::TokenCredentialOptions
+  {
+    /**
+     * @brief Authentication authority URL.
+     * @note Default value is Azure AD global authority (https://login.microsoftonline.com/).
+     *
+     * @note Example of an authority host string: "https://login.microsoftonline.us/". See national
+     * clouds' Azure AD authentication endpoints:
+     * https://docs.microsoft.com/azure/active-directory/develop/authentication-national-cloud.
+     */
+    std::string AuthorityHost = _detail::ClientCredentialCore::AadGlobalAuthority;
+
+    /**
+     * @brief For multi-tenant applications, specifies additional tenants for which the credential
+     * may acquire tokens. Add the wildcard value `"*"` to allow the credential to acquire tokens
+     * for any tenant in which the application is installed.
+     */
+    std::vector<std::string> AdditionallyAllowedTenants;
+  };
+
+  /**
+   * @brief Workload Identity Credential supports Azure workload identity authentication on
+   * Kubernetes and other hosts supporting workload identity. See the Azure Kubernetes Service
+   * documentation at https://learn.microsoft.com/azure/aks/workload-identity-overview for more
+   * information.
+   *
+   */
+  class WorkloadIdentityCredential final : public Core::Credentials::TokenCredential {
+  private:
+    _detail::TokenCache m_tokenCache;
+    _detail::ClientCredentialCore m_clientCredentialCore;
+    std::unique_ptr<_detail::TokenCredentialImpl> m_tokenCredentialImpl;
+    std::string m_requestBody;
+    std::string m_tokenFilePath;
+
+    explicit WorkloadIdentityCredential(
+        std::string tenantId,
+        std::string const& clientId,
+        std::string const& tokenFilePath,
+        std::string const& authorityHost,
+        std::vector<std::string> additionallyAllowedTenants,
+        Core::Credentials::TokenCredentialOptions const& options);
+
+  public:
+    /**
+     * @brief Constructs a Workload Identity Credential.
+     *
+     * @param tenantId Tenant ID.
+     * @param clientId Client ID.
+     * @param tokenFilePath Path of a file containing a Kubernetes service account token.
+     * @param options Options for token retrieval.
+     */
+    explicit WorkloadIdentityCredential(
```

Comment:
Why are there two initializers here, one which takes a TokenCredentialOptions and another which takes a WorkplaceTokenCredentialOptions (which is also a TokenCredentialOptions)?

Reply:
That felt odd to me as well, but I left it for consistency.

This matches all the existing credential types we have (client secret, client certificate, etc.). We have one ctor for the base class TokenCredentialOptions and another for the specific-derived class.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/4874

Reply:
That is SO weird and results in that really awkward "construct a brand new `WorkplaceTokenCredentialOptions` object just to pull the two fields from `WorkplaceTokenCredentialOptions`  (twice!) behavior.